docs/develop-templates.md

```diff
@@ -0,0 +1,54 @@
+# Developing Templates
+
+Unlike fully released [azd templates](https://github.com/topics/azd-templates), the template definition contained in this repository under the `templates` folder is disassembled into multiple directories. This allows re-use of specific assets such as Bicep files, resource assets, tests across templates. To reassemble the assets, a `repoman` file (`repo.yml`) is defined for each template that defines stitching and path rewriting via the `repoman generate` command. To learn more about the `repoman` tooling, see it's source under [generators/repo](../generators/repo).
+
+The `templates` directory in this repository is structured as follows:
+
+```yml
+templates: # Root template folder
+    todo: # Templates related to the Simple ToDo application
+        api: # API servers definition
+        web: # Web servers definition
+        common: # Common assets shared across templates
+        projects: # The actual definition of how to assemble the template projects. repo.yml is defined here which references relevant assets from  different directories. 
+```
+
+## Building
+
+The simplest option is to build all templates in the repository:
+
+```pwsh
+./eng/scripts/Build-Templates.ps1
+```
+
+After this command completes, all generated templates can be found under the `.output` folder. `azd` commands can be ran in any of the generated folders for manual testing.
+
+It is recommended to filter to specific templates for faster build times:
+
+## Build templates with matching name
+
+```pwsh
+./eng/scripts/Build-Templates.ps1 todo-csharp-mongo
+```
+
+Builds the template named `todo-csharp-mongo`.
+
+## Build templates with matching name regex
+
+```pwsh
+./eng/scripts/Build-Templates.ps1 python
+```
+
+Builds the template with names matching the regular expression `python`.
+
+## Build templates under a specific directory
+
+```pwsh
+./eng/scripts/Build-Templates.ps1 -Path ./templates/todo/projects/python-mongo
```

Comment:
is it an option to integrate this logic directly to repoMan? (as we own that tool?)
Wondering about creating this pwsh file on top of the repoMan capabilities vs just adding the feature to repoMan...

The pwsh file script allow quick updates for the generation of templates, but it creates a dependency for repoMan ( we will need to be careful in the future about not introducing breaking changes between them)

Reply:
That's a good question. There are two features this script adds that could belong to `repoman`:
1. A command for auto-discovering nested `repo.yaml`(s) and calling generate on each
1. A flag for preserving a specific folder upon generation to preserve `.env` folder. 

Number 1. could be a desirable feature, although the use-case for having multiple nested templates is not common. The problem here is the integration with "automatic pushing to Git repository", where different templates might have different repository names. The way I think we could solve this is probably to break apart "folder/files stitching" from "repository syncing" as two separate tooling.

Number 2. is a potentially confusing feature outside of `repoman`. I think the generated directory should be 100% idempotent, but we want to preserve `.env` folder for local dev purposes.

Ultimately, whether we should add it to `repoman` should solely depend on whether it is a good feature for the tooling. Breaking changes are not really a good consideration -- if we make a breaking change to `repoman generate` or `repoman list`, obviously all tooling will break in some way. I'm inclined to say no for now.